### PR TITLE
Refactoring and cleaning up

### PR DESCRIPTION
**Pull request by tesa237**
* Introduced interfaces (and initial implementations) and keys (typesafety)
* Added smoke test
* Migrated away from LiveApi

### CODE REVIEW

**Comment by tesa237** [comment]
Sadly the unit test doesn't work because of a date format issue. If anyone knows how to solve it feel free to give me a hint

**Comment by evanova** [comment]
I am going to need more time to review this but one first note: please keep it JDK1.7 compatible.
I am all for Java8 but the library will be hard to cope with using Android.

**Review by evanova — COMMENTED** [review]
I don't think we want those keys. The usage of long ids are very common among all Eve APIs. There is no need for an extra layer.

Also be aware that ids should be long values although the ESI api generation says otherwise. Eve ids are now all 32 bits.

**Review by evanova — COMMENTED** [review]
ESIXXXService interfaces: I am fine with those but I also fine with a Facade containing every call. I think the later would be simpler to manage.

**Review by evanova — COMMENTED** [review]
Business Objects: please keep them as POJO. Interfaces really aren't needed for data transfer objects and it's only going to make it harder on users and implementors.

Likewise, I think builder are a convenience that should be left out of the library and up to the users to have builders. The fact that the new interfaces don't have setters lead to that builder pattern, but both really should not exist. 

What we CAN do is add a utility project that would contain such builders; but keep them away from the core lib.

Edit: builders are good when we need immutable objects that would require calling a constructor with a lot of parameters. They don't serve much purpose on POJOs because POJO don't have to be immutable and have appropriate setters/getters.

**Review by evanova — COMMENTED** [review]
Tests: at this stage I could not care less about tests. It's great if they exist but let's expect that they are broken and live with it for now.

**Review by evanova — COMMENTED** [review]
sourceCompatibility: sadly, part of the reason for this repo is to be used in Android and we cannot have nice Java8 things with Android.

File: client/src/main/java/org/devfleet/esi/api/service/ESIServiceFactory.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import org.devfleet.esi.api.service.ESICharacterService;
+import org.devfleet.esi.impl.CorporationData;
+import org.devfleet.esi.impl.MailboxData;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface ESIServiceFactory {
+
+    Optional<Character> tryFetchCharacter(CharacterKey key);
```
Comment (evanova) [inline client/src/main/java/org/devfleet/esi/api/service/ESIServiceFactory.java:14]: I know Optional is popular but personally I never liked the idea. Returning null is a perfectly viable contract in most cases.

File: client/src/main/java/org/devfleet/esi/impl/service/ESICorporationServiceImpl.java [hunk 2]
```diff
@@ -10,68 +14,73 @@
...
-                    .body();
-            final List<Corporation.Member> r = new ArrayList<>(members.size());
+                api
+                        .getCorporationsCorporationIdMembers(key.getId(), datasource)
+                        .execute()
+                        .body();
+            final List<CorporationMember> r = new ArrayList<>(members.size());
             for (GetCorporationsCorporationIdMembers200Ok m: members) {
-                r.add(EsiTransformer.transform(m));
+                r.add(ESITransformer.transform(m));
```
Comment (evanova) [inline client/src/main/java/org/devfleet/esi/impl/service/ESICorporationServiceImpl.java:68]: I can't believe there is someone out there who is more adamant on proper usage of capitalization in abbreviations.

+1 for ESI :)

File: client/src/main/java/org/devfleet/esi/impl/service/ESICorporationServiceImpl.java [hunk 3]
```diff
@@ -10,68 +14,73 @@
 
 import java.io.IOException;
 import java.util.ArrayList;
-import java.util.Collections;
 import java.util.List;
+import java.util.Optional;
 
-class CorporationAPIImpl {
+import static java.util.Collections.emptyList;
```
Comment (evanova) [inline client/src/main/java/org/devfleet/esi/impl/service/ESICorporationServiceImpl.java:20]: Don't do static imports please. They are hard to read and follow.

File: client/src/main/java/org/devfleet/esi/impl/service/ESICorporationServiceImpl.java [hunk 4]
```diff
@@ -10,68 +14,73 @@
...
 
-    private static Logger LOG = LoggerFactory.getLogger(CorporationAPIImpl.class);
```
Comment (evanova) [inline client/src/main/java/org/devfleet/esi/impl/service/ESICorporationServiceImpl.java:18]: should also be final

File: client/src/main/java/org/devfleet/esi/api/service/ESIServiceFactory.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package org.devfleet.esi.api.service;
```
Comment (evanova) [inline client/src/main/java/org/devfleet/esi/api/service/ESIServiceFactory.java:1]: So I was mentioning a facade in a previous comment. 
This interface definitely is one and it's great. Should it be simply called ESI?

File: client/src/main/java/org/devfleet/esi/impl/service/ESICorporationServiceImpl.java (hunk 3 above)
Comment (tesa237) [reply client/src/main/java/org/devfleet/esi/impl/service/ESICorporationServiceImpl.java:20]: Fair enough. I'm used to them and don't see much of a problem with them, but they can take some time getting used to. 

Will revert to "normal" imports

File: client/src/main/java/org/devfleet/esi/impl/service/ESICorporationServiceImpl.java (hunk 4 above)
Comment (tesa237) [reply client/src/main/java/org/devfleet/esi/impl/service/ESICorporationServiceImpl.java:18]: agreed

File: client/src/main/java/org/devfleet/esi/api/service/ESIServiceFactory.java (hunk 5 above)
Comment (tesa237) [reply client/src/main/java/org/devfleet/esi/api/service/ESIServiceFactory.java:1]: Sure, why not.

**Comment by tesa237** [comment]
The keys are for type safety. In my experience it will pay off to have Key objects. Just having a long/int as identifier will sooner or later bite you - you'll end up fetching a Character with a Corporation identifier. With keys this is almost unlikely to happen. 

As for the Optionals, with java 7 that'll fall away. I'll keep the name "tryFetch" though, it indicates that things CAN go wrong and developers should keep that in mind while using the API

The ESIXXXServices can, in my larger vision, be used either on their own or via a facade. I don't see anything cancelling either view.

With the immutable DataObjects, I find that is good practice to keep data objects immutable. Especially things that are fetch from the API that are intended to be read-only should stay immutable. If there is a setter developers will use it. And just having a setter because we're lazy and don't want to implement a builder is just poor a argument. If a developer shouldn't call a method I prefer not even having the method at all. 
One can argue that we're talking about sensible people here etc etc, but keeping things lean and clear will pay off in the end.

**Comment by evanova** [comment]
I have removed the LiveAPI from the master. 

After another review of this PR, I will not merge it. I really do not think a Key class is needed and definitely do not want the overhead of model objects being represented by interfaces. Immutability is not a concern at this layer in my view and can be introduced by higher level APIs should clients of this library choose to do this.

**Comment by evanova** [comment]
Newer API

**Pull request closed without merging**